tests/lib/in_memory/database.py

```diff
@@ -0,0 +1,253 @@
+"""An implementation of a simple and fast in-memory database that can be driven by the
+the in-memory engine.
+
+There are two classes:
+
+    * Tables contain a mapping from a name to a column.
+    * Columns contain a mapping from a patient id to a list of values, and have a
+      default value (which defaults to None) for any patients not in the mapping.
+
+In most cases, this database does not care whether the table or column contains one row
+per patient or many rows per patient, and instead relies on the in-memory engine doing
+the right thing with a valid QM object.
+
+There are two places where this database does care:
+
+    * When applying a binary operation on two columns, if column C1 contains one row
+      per patient and column C2 contains many rows per patient, we have to create
+      a mapping with values from C1 with the shape of C2.
+    * When getting the single value from a column with one row per patient, we verify
+      that the column does contain one row per patient.
+"""
```

Comment:
I think it's a good idea to have this engine. We should promise ourselves that we'll keep an eye on the cost of maintaining it.

I suspect that there are some hard-won insights that might make bits of the implementation obscure to non-Peters in the future, but the thorough testing mitigates that.

Reply:
> I suspect that there are some hard-won insights that might make bits of the implementation obscure to non-Peters in the future, but the thorough testing mitigates that.

A non-Peter should do the work to extend it to cover the remaining parts of the QM.  The QM nodes in question have visitors with `assert False` in the body.

Reply:
I strongly agree.